### PR TITLE
Automatic unpacking of URL instances

### PR DESCRIPTION
**Pull request by duylethanh**
This PR contains the following:
- support automatic unpacking of URL isntances via a new ``format`` key
- documentation on automatic unpacking of URL instances
- tests for automatic unpacking of URL instances
- add some missing keys in [experiments.yml reference page](https://simexpal.readthedocs.io/en/latest/experiments_reference.html)

### CODE REVIEW

File: simexpal/instances.py [hunk 1]
```diff
@@ -98,8 +99,31 @@ def substitute(p):
...
+				try:
+					shutil.unpack_archive(download_path, extract_dir=instances_dir, format=inst_yml['format'])
+					os.rename(os.path.join(instances_dir, filename), partial_path + ext)
+				except (ValueError, shutil.ReadError):
+					if inst_yml['format'] in ADDITIONAL_SUPPORTED_ARCHIVE_FORMATS:
+						if inst_yml['format'] == 'gz':
+							with gzip.open(download_path, 'rb') as f_in:
+								with open(tmp_path, 'wb') as f_out:
+									shutil.copyfileobj(f_in, f_out)
+							os.rename(tmp_path, partial_path + ext)
```
Comment (avdgrinten) [inline simexpal/instances.py:118]: I don't like the fact that we first try one unpacking function and then fall back to a different one. What's the reason to do that? Why don't we check against `ADDITIONAL_SUPPORTED_ARCHIVE_FORMATS` first and fall back to `shutil.unpack_archive` if we do not find the format in the list?

File: simexpal/instances.py [hunk 2]
```diff
@@ -98,8 +99,31 @@ def substitute(p):
...
-			download_path = partial_path + ext
+			download_path = os.path.join(instances_dir, filename + '.download')
 			download_as_stream(url, download_path)
+
+			if inst_yml.get('format', None):
+				import shutil
+
+				tmp_path = os.path.join(instances_dir, filename + '.tmp')
+				try:
+					shutil.unpack_archive(download_path, extract_dir=instances_dir, format=inst_yml['format'])
```
Comment (avdgrinten) [inline simexpal/instances.py:110]: I think we should extract to a temporary subdirectory of instances_dir here. The zip file might contain all kinds of files that clash with other instances.

Comment (avdgrinten) [reply simexpal/instances.py:110]: We can use `os.rename` afterwards to move the right file to the instances_dir, and remove the temporary directory.

**Review by avdgrinten — COMMENTED** [review]
Some comment below. A more general question is: how do we control which file we extract from the archive (in case the archive contains multiple files)?

File: simexpal/instances.py (hunk 1 above)
Comment (duylethanh) [reply simexpal/instances.py:118]: So we should do something like the following?:
``` python
#key: format; value: function for unpacking 
ADDITIONAL_SUPPORTED_ARCHIVE_FORMATS = {"gz": func_to_extract_gz}  format
```

``` python

if inst_yml['format'] in ADDITIONAL_SUPPORTED_ARCHIVE_FORMATS:
    # use function at ADDITIONAL_SUPPORTED_ARCHIVE_FORMATS[inst_yml['format']] to unpack
else:
    try:
        shutil.unpack_archive(...)
    except appropriate_errors:
         # skip unpacking of this instance
```

File: simexpal/instances.py (hunk 2 above)
Comment (duylethanh) [reply simexpal/instances.py:110]: I agree

**Comment by duylethanh** [comment]
> 
> 
> Some comment below. A more general question is: how do we control which file we extract from the archive (in case the archive contains multiple files)?

Maybe we could allow the following syntax motivated by the syntax of [arbitrary input files](https://simexpal.readthedocs.io/en/latest/instances.html#arbitrary-input-files):

``` yaml
instances:
     - method: url
       url: https://some_url/@INSTANCE_FILENAME@.txt.gz
       items:
         - 'bar'
         - name: foo
            files:
              - file1
              - file2
       format: gz       
 ```
 
 In the "normal" case where we just have one file we subtitute ``bar`` into ``@INSTANCE_FILENAME@``, download from that URL and unpack it (according to ``format``) in a temp dir and move the respective file .

In the second case, where we have multiple files we substitute  ``foo`` into ``@INSTANCE_FILENAME@``, download from that URL and unpack it in a temp dir and move the respective files from the ``files`` list.

File: simexpal/instances.py (hunk 1 above)
Comment (avdgrinten) [reply simexpal/instances.py:118]: Yes, that sounds better.

Comment (duylethanh) [reply simexpal/instances.py]: I changed it according to above, but did not include the unpacking function into ``ADDITIONAL_SUPPORTED_ARCHIVE_FORMATS``.

File: simexpal/instances.py (hunk 2 above)
Comment (duylethanh) [reply simexpal/instances.py]: Done.

**Comment by duylethanh** [comment]
> > Some comment below. A more general question is: how do we control which file we extract from the archive (in case the archive contains multiple files)?
> 
> Maybe we could allow the following syntax motivated by the syntax of [arbitrary input files](https://simexpal.readthedocs.io/en/latest/instances.html#arbitrary-input-files):
> 
> ```yaml
> instances:
>      - method: url
>        url: https://some_url/@INSTANCE_FILENAME@.txt.gz
>        items:
>          - 'bar'
>          - name: foo
>             files:
>               - file1
>               - file2
>        format: gz       
> ```
> 
> In the "normal" case where we just have one file we subtitute `bar` into `@INSTANCE_FILENAME@`, download from that URL and unpack it (according to `format`) in a temp dir and move the respective file .
> 
> In the second case, where we have multiple files we substitute `foo` into `@INSTANCE_FILENAME@`, download from that URL and unpack it in a temp dir and move the respective files from the `files` list.

As far as I know, jsonschema can not validate such schemes (where the array has elements of different types at the same time - except if we set the ``items`` key to an arbitrary array without validating its elements, but I'm not sure thats the way to go), so we can dismiss this approach for now.

It does not matter if the archive contains multiple files, if we only want to extract one.
If we want to support the extraction of multiple files, we might be able to support it if we only allow dicts in the ``items`` key when the ``format`` key is present, e.g.:

 ```yaml
 instances:
      - method: url
        url: https://some_url/@INSTANCE_FILENAME@.zip
        items:
          - name: foo
             files:
               - file1
               - file2
           - name: bar
             files:
               - file3
               - file4
        format: zip       
 ```

where ``file1`` and ``file2`` will be extracted from ``foo.zip`` and analogously for ``bar``.

**Comment by duylethanh** [comment]
PR ready to be reviewed

**Comment by avdgrinten** [comment]
To clarify: if the archive is called `Bar.zip` (and stored at a location `https://example.com/Bar.zip`), how would I make simexpal extract file `Foo` out of `Bar.zip`?

**Comment by duylethanh** [comment]
Seems like I did not implement it at that time , but it would probably be like in the comment [above](https://github.com/hu-macsy/simexpal/pull/114#issuecomment-1013743476). I will implement it after I'm done with refining the other open PRs.

**Comment by bernlu** [comment]
Downloading instances is no longer planned for future releases.

**Pull request closed without merging**